component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java

```diff
@@ -0,0 +1,281 @@
+        moesifKeyClientMap = new ConcurrentHashMap();
+    }
+
+    public static synchronized MoesifKeyRetriever getInstance(String authUsername, String authPwd) {
+        if (moesifKeyRetriever == null) {
+            return new MoesifKeyRetriever(authUsername, authPwd);
+        }
+        return moesifKeyRetriever;
+    }
+
+    /**
+     * Will initialize the empty orgID-MoesifKey map.
+     * Will refresh/refill the  orgID-MoesifKey map.
+     */
+    public void initOrRefreshOrgIDMoesifKeyMap() {
+        int attempts = MoesifMicroserviceConstants.NUM_RETRY_ATTEMPTS;
+        try {
+            callListResource();
+        } catch (IOException | APICallException ex) {
+            // TODO: Separate retry logic to a separate class.
+            log.error("First attempt failed,retrying.", ex);
+            while (attempts > 0) {
+                attempts--;
+                try {
+                    Thread.sleep(MoesifMicroserviceConstants.TIME_TO_WAIT);
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+                try {
+                    callListResource();
+                } catch (IOException | APICallException e) {
+                    log.error("Retry attempt failed.", e);
+                }
+            }
+        }
+    }
+
+    /**
+     * Will retrieve single Moesif key corresponding to given organization ID.
+     *
+     * @param orgID
+     * @return Moesif Key corresponding orgID
+     */
+    public String getMoesifKey(String orgID) {
+        String response;
+        int attempts = MoesifMicroserviceConstants.NUM_RETRY_ATTEMPTS;
+        try {
+            response = callDetailResource(orgID);
+        } catch (IOException | APICallException ex) {
+            // TODO: Separate retry logic to a separate class.
+            log.error("First attempt failed,retrying.", ex);
+            while (attempts > 0) {
+                attempts--;
+                try {
+                    Thread.sleep(MoesifMicroserviceConstants.TIME_TO_WAIT);
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+                try {
+                    response = callDetailResource(orgID);
+                    return response;
+                } catch (IOException | APICallException e) {
+                    log.error("Retry attempt failed.", e);
+                }
+            }
+            response = null;
+        }
+        return response;
+    }
+
+    /**
+     * Will remove Moesif key corresponding to organization ID.
+     * Existing Moesif SDK client associated with the moesif key will be also removed.
+     *
+     * @param orgID
+     */
+    public void removeMoesifKeyFromMap(String orgID) {
+        String moesifKey = orgIDMoesifKeyMap.remove(orgID);
+        moesifKeyClientMap.remove(moesifKey);
+    }
+
+    private void callListResource() throws IOException, APICallException {
```

Comment:
Shall we add a proper doc comment for this method?

Reply:
Ack,